File: RegexInterface.java
public interface RegexInterface {
        //(* A list of uppercase and lowercase letters. *)
        String letter = "(A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|a|b|c|d|e|f|g|h|i|j|k|l|m|n|o|p|q|r|s|t|u|v|w|x|y|z)";

        //(* Rules for all numbers. *)
        String sign = "((\\+|\\-)?)";

        //(* Rules for binary numbers. *)
        String binary_digit = "(0|1)";
        String binary_prefix = "(0b)";
        String binary_number = "(" + sign + binary_prefix + "(" + binary_digit + "+?))";

        //(* Rules for octal numbers. *)
        String octal_digit = "(" + binary_digit + "|2|3|4|5|6|7)";
        String octal_prefix = "(0o)";
        String octal_number = "(" + sign + octal_prefix + "(" + octal_digit + "+?))";

        //(* Rules for decimal numbers. *)
        String decimal_digit = "(" + octal_digit + "|8|9)";
        String decimal_number = "(" + sign + "(" + decimal_digit + "+?))";

        //(* Rules for hexidecimal numbers. *)
        String hex_digit = "(" + decimal_digit + "|A|B|C|D|E|F|a|b|c|d|e|f)";
        String hex_prefix = "(0x)";
        String hex_number = "(" + sign + hex_prefix + "(" + hex_digit + "+?))";

        //(* Rules for string literal characters. *)
        String string_chars_double_quote = "((?:[^\\\"]|\\\")*)";
        String string_chars_single_quote = "((?:[^\\']|\\')*)";

        //(* Rules defining basic type literals. *)
        String int_literal = "(" + binary_number + "|" + octal_number + "|" + decimal_number + "|" + hex_number + ")";
        String real_literal = "(" + decimal_number + "." + decimal_number + ")";
        String bool_literal = "true|false";
        String string_literal = "(('" + string_chars_single_quote + "')" + "|(\"string_chars_double_quote\"))";
        String literal_value = "(" + int_literal + "|" + real_literal + "|" + bool_literal + "|" + string_literal + ")";

        //(* Rules for variables *)
        String var_name = "(" + letter + "((" + letter + "|" + decimal_digit + "|_)*))";
        String var_or_literal = "(" + var_name + "|" + literal_value + ")";
        String var_or_int = "(" + var_name + "|" + int_literal + ")";
        String var_type = "(int|real|string|bool)";

        
        //(* Rules for operators *)
        String operator_1 = "(" + sign + "|" + "\\|\\|)";
        String  operator_2 = "(*|\\/|%|&&)";
        String operator_3 = "(\\*\\*)";
        String operator_comparison = "(==|!=|<|<=|>=|>)";
        
        //(* Rules for function declarations *)
        String parameter = "(" + var_type + var_name + ")";
        String default_parameter = "(" + parameter + "=" + literal_value + ")";
        String parameter_list = "(" + "((" + parameter + "((," + parameter + ")*)" + "((," + default_parameter + ")*)" + ")?)" + "|" + "((" + default_parameter + "((," + default_parameter + ")*)" + ")?))";
}

File: NajimJS/RegexInterface.js
// This is an interface to hold all regular expressions.
var RegexInterface;
(function (RegexInterface) {
    RegexInterface.letter = "(A|B|C|D|E|F|G|H|I|J|K|L|M|N|O|P|Q|R|S|T|U|V|W|X|Y|Z|a|b|c|d|e|f|g|h|i|j|k|l|m|n|o|p|q|r|s|t|u|v|w|x|y|z)";
    RegexInterface.sign = "((\\+|\\-)?)";
    RegexInterface.binary_digit = "(0|1)";
    RegexInterface.binary_prefix = "(0b)";
    RegexInterface.binary_number = "(" + RegexInterface.sign + RegexInterface.binary_prefix + "(" + RegexInterface.binary_digit + "+?))";
    RegexInterface.octal_digit = "(" + RegexInterface.binary_digit + "|2|3|4|5|6|7)";
    RegexInterface.octal_prefix = "(0o)";
    RegexInterface.octal_number = "(" + RegexInterface.sign + RegexInterface.octal_prefix + "(" + RegexInterface.octal_digit + "+?))";
    RegexInterface.decimal_digit = "(" + RegexInterface.octal_digit + "|8|9)";
    RegexInterface.decimal_number = "(" + RegexInterface.sign + "(" + RegexInterface.decimal_digit + "+?))";
    RegexInterface.hex_digit = "(" + RegexInterface.decimal_digit + "|A|B|C|D|E|F|a|b|c|d|e|f)";
    RegexInterface.hex_prefix = "(0x)";
    RegexInterface.hex_number = "(" + RegexInterface.sign + RegexInterface.hex_prefix + "(" + RegexInterface.hex_digit + "+?))";
    RegexInterface.string_chars_double_quote = "((?:[^\\\"]|\\\")*)";
    RegexInterface.string_chars_single_quote = "((?:[^\\\']|\\\')*)";
    RegexInterface.int_literal = "(" + RegexInterface.binary_number + "|" + RegexInterface.octal_number + "|" + RegexInterface.decimal_number + "|" + RegexInterface.hex_number + ")";
    RegexInterface.real_literal = "(" + RegexInterface.decimal_number + "." + RegexInterface.decimal_number + ")";
    RegexInterface.bool_literal = "true|false";
    RegexInterface.string_literal = "((\'" + RegexInterface.string_chars_single_quote + "\')" + "|(\"" + RegexInterface.string_chars_double_quote + "\"))";
    RegexInterface.literal_value = "(" + RegexInterface.int_literal + "|" + RegexInterface.real_literal + "|" + RegexInterface.bool_literal + "|" + RegexInterface.string_literal + ")";
    RegexInterface.var_name = "(" + RegexInterface.letter + "((" + RegexInterface.letter + "|" + RegexInterface.decimal_digit + "|_)*))";
    RegexInterface.var_or_literal = "(" + RegexInterface.var_name + "|" + RegexInterface.literal_value + ")";
    RegexInterface.var_or_int = "(" + RegexInterface.var_name + "|" + RegexInterface.int_literal + ")";
    RegexInterface.var_type = "(int|real|string|bool)";
    RegexInterface.operator_1 = "(" + RegexInterface.sign + "|" + "\\|\\|)";
    RegexInterface.operator_2 = "(*|\\/|%|&&)";
    RegexInterface.operator_3 = "(\\*\\*)";
    RegexInterface.operator_comparison = "(==|!=|<|<=|>=|>)";
    RegexInterface.parameter = "(" + RegexInterface.var_type + RegexInterface.var_name + ")";
    RegexInterface.default_parameter = "(" + RegexInterface.parameter + "=" + RegexInterface.literal_value + ")";
    RegexInterface.parameter_list = "(" + "((" + RegexInterface.parameter + "((," + RegexInterface.parameter + ")*)" + "((," + RegexInterface.default_parameter + ")*)" + ")?)" + "|" + "((" + RegexInterface.default_parameter + "((," + RegexInterface.default_parameter + ")*)" + ")?))";

    RegexInterface.statement_block = "(" + "\{" + "(" + RegexInterface.statement + ")*" + "\})";
    RegexInterface.statement = "(" + RegexInterface.statement_flow +"|"+ "(" + RegexInterface.statement_function_call +"|"+ RegexInterface.statement_var_declare +"|"+ RegexInterface.statement_var_assign +"|" + RegexInterface.statement_comment + ")" + ";)";
    RegexInterface.statement_function_call = "(" + RegexInterface.var_name + "\(" + "(" + RegexInterface.var_or_literal + "("+","+ RegexInterface.var_or_literal + ")*" + ")?" + "\))";
    RegexInterface.statement_var_declare = "(" + "(const)?" + RegexInterface.var_type + "=" + RegexInterface.expression + ")";
    RegexInterface.statement_var_assign = "("+RegexInterface.var_name + "("+"=" +"|"+ "\+=" +"|"+ "\-=" +"|"+ "\*=" +"|"+ "\/=" +"|"+ "\%=" +"|"+ "\*\*="+")"+ RegexInterface.expression+")";
    RegexInterface.statement_comment = RegexInterface.comment;
    RegexInterface.statement_flow = "("+RegexInterface.statement_if + "|"+ RegexInterface.statement_for +"|"+ RegexInterface.statement_while +")";

    RegexInterface.statement_if = "("+"if"+ "\("+ RegexInterface.expression+ "\)"+ RegexInterface.statement_block+ "("+ "else" + RegexInterface.statement_block + ")?)";
    RegexInterface.statement_for = "(" + "for" + "\(" + "("+RegexInterface.var_name+ "="+")?" + RegexInterface.var_or_int + "("+"to"+ "|"+ "downto"+")"+ RegexInterface.var_or_int+ "("+"step"+ RegexInterface.var_or_int+")?"+ "\)"+ RegexInterface.statement_block + ")";
    RegexInterface.statement_while = "("+"while"+ "\("+ RegexInterface.expression+ "\)"+ RegexInterface.statement_block + ")";

    RegexInterface.expression = "(" + RegexInterface.expression_1 + "(" + RegexInterface.operator_comparison + RegexInterface.expression_1 + ")?)";
    RegexInterface.expression_1 = "(("+ RegexInterface.sign +")?" + RegexInterface.expression_2 + "("+RegexInterface.operator_1+ RegexInterface.expression_2+")*)";
    RegexInterface.expression_2 = "("+RegexInterface.expression_3+ "(" + RegexInterface.operator_2 + RegexInterface.expression_3 + ")*)";
    RegexInterface.expression_3 = "("+RegexInterface.var_or_literal+ "|"+ "\("+ RegexInterface.expression+ "\)" + "|"+ "\!" + RegexInterface.expression_3 + "|"+ RegexInterface.expression_3 + RegexInterface.operator_3 + RegexInterface.expression3 + ")";

    RegexInterface.function_def = "("+"function"+ RegexInterface.var_name+ "\("+ RegexInterface.parameter_list + "\)"+ RegexInterface.statement_block + ")";

})(RegexInterface || (RegexInterface = {}));

exports.RegexInterface = RegexInterface;